presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1586,7 +1593,167 @@ public static Double stSphericalLength(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) S
+        }
+
+        if (containsEitherPole(polygon)) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "When applied to SphericalGeography inputs, ST_Contains only supports polygons which do not enclose poles.");
+        }
+
+        int k = getNumberIntersectionsWithPolygon(polygon, point);
+        if (k % 2 == 1) {
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    private static boolean liesInRange(double point, double rangeStart, double rangeEnd)
+    {
+        return point >= rangeStart && point <= rangeEnd;
+    }
+
+    private static boolean edgeCrosses(double edgeStart, double edgeEnd, double longitude)
+    {
+        double start = min(edgeStart, edgeEnd);
+        double end = max(edgeStart, edgeEnd);
+        double distanceVia0 = end - start;
+        double distanceVia180 = start - MIN_LONGITUDE + MAX_LONGITUDE - end;
+        if (distanceVia0 < distanceVia180) {
+            return liesInRange(longitude, start, end);
+        }
+        else {
+            return liesInRange(longitude, MIN_LONGITUDE, start) || liesInRange(longitude, end, MAX_LONGITUDE);
+        }
+    }
+
+    private static boolean edgeCrossesLongitude(double edgeStart, double edgeEnd, double longitude)
+    {
+        double start = min(edgeStart, edgeEnd);
+        double end = max(edgeStart, edgeEnd);
+        double distanceVia0 = end - start;
+        double distanceVia180 = start - MIN_LONGITUDE + MAX_LONGITUDE - end;
+        if (distanceVia0 < distanceVia180) {
+            return liesInRange(longitude, start, end);
+        }
+        else {
+            return liesInRange(longitude, MIN_LONGITUDE, start) || liesInRange(longitude, end, MAX_LONGITUDE);
+        }
+    }
+
+    private static boolean edgeIntersectsTestEdge(Point arcStart, Point arcEnd, Point longitudeStart, Point longitudeEnd)
+    {
+        // y of longstart must be lower than y of longend
+        double longitude = longitudeStart.getX();
+        boolean arcCrossesLongitude = edgeCrossesLongitude(arcStart.getX(), arcEnd.getX(), longitude);
+
+        if (arcCrossesLongitude) {
+            double edgeLatitude = getArcLatitudeAtLongitude(arcStart, arcEnd, longitude);
+            boolean inRange = liesInRange(edgeLatitude, longitudeStart.getY() - 0.5, longitudeEnd.getY() + 0.5);
+            return inRange;
+        }
+        else {
+            return false;
+        }
+    }
+
+    private static double getArcLatitudeAtLongitude(Point start, Point end, double longitude)
+    {
+        SphericalExcessCalculator calculator = new SphericalExcessCalculator(start);
+        return calculator.computeExtrapolatedLatitudeThroughPoint(end, longitude);
+    }
+
+    private static Point getExternalPoint(double longitude)
+    {
+        return new Point(longitude, 90.0);
+    }
+
+    private static int getNumberIntersectionsWithPolygon(Polygon polygon, Point point)
```

Comment:
You should consider a signed sum here, which is more resilient to invalid polygons and other edge cases:
http://geomalgorithms.com/a03-_inclusion.html